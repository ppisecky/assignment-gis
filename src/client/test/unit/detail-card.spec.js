import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import DetailCard from '../../components/detail-card'
import ElementUI from 'element-ui';
import { places } from './__mocks__/mockedData'
import flushPromises from 'flush-promises'
import { randomPlace, randomSquare } from './helpers'

const localVue = createLocalVue();
localVue.use(ElementUI);

const mockAxios = { $get: () => Promise.resolve({}) }
const shallowMountDetailCard = (settings = {}) => {
    const defaultSettings = {
        localVue, 
        mocks: { $axios: mockAxios },
        propsData: {
            entity: randomPlace(),
            isSquare: false
        }
    }
    return shallowMount(DetailCard, { ...defaultSettings, ...settings })
}

const freshMountDetailCard = (settings = {}) => {
    const freshLocalVue = createLocalVue();
    freshLocalVue.use(ElementUI);

    const mocks = { $axios: mockAxios }
    return mount(DetailCard, { ...{ localVue: freshLocalVue, mocks }, ...settings })
}

describe('detail card', () => {
    it('mounts properly', () => {
        // given

        // when
        const testSubject = shallowMountDetailCard()

        // then
        expect(testSubject.isVueInstance()).toBeTruthy()
    })

    it('has a mounted hook', () => {
        // given

        // when
        const testSubject = DetailCard

        // then
        expect(typeof testSubject.mounted).toBe('function')
    })

    it('initializes with correct data', () => {
        // given
        const desiredPropertyBlacklist = ['way', 'z_order', 'name']

        // when
        const testSubject = DetailCard
        
        // then
        expect(typeof testSubject.data).toBe('function')
        const defaultData = testSubject.data()
        expect(defaultData.loadingDetail).toBe(false)
        expect(defaultData.propertiesBlacklist).toEqual(expect.arrayContaining(desiredPropertyBlacklist))
    })

    it('emits event on close button click', () => {
        // given
        const propsData = {
            entity: randomPlace(),
            isSquare: false
        }

        // when
        const testSubject = mount(DetailCard, {
            localVue,
            mocks: {$axios: mockAxios},
            propsData
        })

        // then
        expect(testSubject.contains('.close-button')).toBe(true)
        const closeButton = testSubject.find('.close-button')
        closeButton.trigger('click')
        expect(testSubject.emitted()['close-detail']).toBeTruthy()
        expect(testSubject.emitted()['close-detail']).toHaveLength(1)
    })

    it('loads fresh data upon mounting', async () => {
        // given
        const place =  randomPlace()
        const propsData = {
            entity: {
                osm_id: place.osm_id,
                name: place.name,
                tourism: place.tourism
            },
            isSquare: false
        }
        const mocks = {
            $axios: {$get: (p) => Promise.resolve(p.includes('places/'+place.osm_id) ? place : randomSquare())}
        }

        // when
        const testSubject = shallowMountDetailCard({
            propsData,
            mocks
        })
        await flushPromises()

        // then
        expect(testSubject.vm.entityDetail).toBe(place)
    })

    it('loads fresh data upon entity change', async () => {
        // given
        let square =  randomSquare()
        const propsData = {
            entity: {
                osm_id: square.osm_id,
                name: square.name,
                tourism: square.tourism
            },
            isSquare: true
        }
        const mocks = {
            $axios: {$get: (p) => Promise.resolve(p.includes('areas/' + square.osm_id) ? square : randomPlace())}
        }

        // when
        const testSubject = shallowMountDetailCard({
            propsData,
            mocks
        })
        await flushPromises()
        square = randomSquare()
        const changedPropsData = {
            entity: {
                osm_id: square.osm_id,
                name: square.name,
                tourism: square.tourism
            }
        }
        testSubject.setProps(changedPropsData);
        await flushPromises()

        // then
        expect(testSubject.vm.entityDetail).toBe(square)
    })

    it('renders correctly', () => {
        // given
        const props = {
            entity: { "osm_id": "4833390172", "name": "Sina Centurion Palace", "geojson": { "type": "Point", "coordinates": [12.3336169, 45.4309549] }, "tourism": "hotel", "lng": 12.3336169, "lat": 45.4309549 },
            isSquare: false
        }

        // when
        const testSubject = freshMountDetailCard({
            propsData: props
        })

        // then
        expect(testSubject).toMatchSnapshot();
    })
})