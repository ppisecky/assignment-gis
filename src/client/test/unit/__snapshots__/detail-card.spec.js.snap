// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`detail card renders correctly 1`] = `
<div class="el-card box-card is-always-shadow" id="detail-card" style="position: relative;">
  <div class="el-card__header">
    <div class="clearfix">
      <h3>
        Sina Centurion Palace<br> <small>Hotel</small> <button type="button" class="el-button close-button el-button--default el-button--mini is-circle">
          <!----><i class="el-icon-close"></i>
          <!----></button></h3>
    </div>
  </div>
  <div class="el-card__body">
    <div class="el-table el-table--fit el-table--fluid-height el-table--scrollable-x" style="display: none;">
      <div class="hidden-columns">
        <div></div>
        <div></div>
      </div>
      <div class="el-table__header-wrapper">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 250px;">
          <colgroup>
            <col name="el-table_2_column_3" width="100">
            <col name="el-table_2_column_4" width="150">
          </colgroup>
          <thead class="">
            <tr class="">
              <th colspan="1" rowspan="1" class="el-table_2_column_3     is-hidden is-leaf">
                <div class="cell">Property</div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_2_column_4     is-leaf">
                <div class="cell">Value</div>
              </th>
            </tr>
          </thead>
        </table>
      </div>
      <div class="el-table__body-wrapper is-scrolling-left" style="max-height: 112px;">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 250px;">
          <colgroup>
            <col name="el-table_2_column_3" width="100">
            <col name="el-table_2_column_4" width="150">
          </colgroup>
          <tbody>
            <!---->
          </tbody>
        </table>
        <div class="el-table__empty-block" style="width: 250px;"><span class="el-table__empty-text">暂无数据</span></div>
        <!---->
      </div>
      <!---->
      <div class="el-table__fixed" style="width: 100px;">
        <div class="el-table__fixed-header-wrapper">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 250px;">
            <colgroup>
              <col name="el-table_2_column_3" width="100">
              <col name="el-table_2_column_4" width="150">
            </colgroup>
            <thead class="">
              <tr class="">
                <th colspan="1" rowspan="1" class="el-table_2_column_3     is-leaf">
                  <div class="cell">Property</div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_2_column_4     is-hidden is-leaf">
                  <div class="cell">Value</div>
                </th>
              </tr>
            </thead>
          </table>
        </div>
        <div class="el-table__fixed-body-wrapper" style="top: 44px; max-height: 112px;">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 250px;">
            <colgroup>
              <col name="el-table_2_column_3" width="100">
              <col name="el-table_2_column_4" width="150">
            </colgroup>
            <tbody>
              <!---->
            </tbody>
          </table>
          <!---->
        </div>
        <!---->
      </div>
      <!---->
      <!---->
      <div class="el-table__column-resize-proxy" style="display: none;"></div>
    </div>
  </div>
</div>
`;
